### PR TITLE
Add more training models and RLHF algorithms

### LINKED ISSUE #6367: [FEATURE]: Add more training models and RLHF algorithms
### Describe the feature

Add more training models and RLHF algorithms for the branch `grpo-latest`.

### PR DESCRIPTION
**Pull request by sglucas**
Closes #6367 

## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Add more training models (LLaMA3, Qwen3) and RLHF algorithms (REINFORCE++, RLOO).  

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 1]
```diff
@@ -180,23 +182,72 @@ def step(self, step_idx: int, pbar: Any, **kwargs) -> Optional[float]:
...
-        advantages = ((reward - reward_mean) / (reward_std + 1e-4)).unsqueeze(dim=-1)
+            advantages = (
+                reward * self.num_generations / (self.num_generations - 1)
+                - reward_mean * self.num_generations / (self.num_generations - 1)
+            ).unsqueeze(dim=-1)
 
-        # [minibatch_size x num_of_generation]
-        loss_mask = torch.ones(action_mask.size(0), device=action_mask.device).bool()
+            # [minibatch_size x num_of_generation]
+            loss_mask = torch.ones(action_mask.size(0), device=action_mask.device).bool()
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/grpo_consumer.py:251]: may be better to move the common calculations outside of the if statements for conciseness

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 2]
```diff
@@ -180,23 +182,72 @@ def step(self, step_idx: int, pbar: Any, **kwargs) -> Optional[float]:
...
+
+            group_reward = reward.view(-1, self.num_generations)
+            reward_mean = group_reward.mean(dim=1)
+            # [minibatch_size x num_generations]
+            reward_mean = reward_mean.repeat_interleave(self.num_generations, dim=0)
+
+            # [minibatch_size x num_generations]
+            advantages = ((reward - reward_mean)).unsqueeze(dim=-1)
+
+            advantages_mean = advantages.mean(dim=0)
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/grpo_consumer.py:223]: Isn't the advantages_mean always 0 as advantage is already zero-centered in the previous step?

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 3]
```diff
@@ -180,23 +182,72 @@ def step(self, step_idx: int, pbar: Any, **kwargs) -> Optional[float]:
...
+
+            advantages_std = advantages.std(dim=0)
+
+            advantages = (advantages - advantages_mean) / (advantages_std + 1e-4)
+
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/grpo_consumer.py:228]: maybe consider double-checking the reinforce++ baseline advantage calculation. In reinforce ++, each sample's advantage is calculated by subtracting the mean reward of all generation in the global batch, not per prompt mean

File: applications/ColossalChat/coati/distributed/untitled.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+4.51.0: qwen2.5 + grpo, qwen3 + grpo, cannot: llama2, llama3.2
+4.47.0:
```
Comment (YeAnbang) [inline applications/ColossalChat/coati/distributed/untitled.txt:2]: remove test log file

File: applications/ColossalChat/rl_example.py [hunk 5]
```diff
@@ -227,13 +227,13 @@
     os.environ["TOKENIZERS_PARALLELISM"] = "false"  # Disable tokenizers parallelism to avoid deadlock
 
     inference_model_config = dict(path=args.model)
-    train_model_config = dict(path=args.model, use_flash_attention_2=True, use_cache=False)
+    train_model_config = dict(path=args.model, use_flash_attention_2=False, use_cache=False)
```
Comment (YeAnbang) [inline applications/ColossalChat/rl_example.py:230]: why is flash attention not supported?

File: applications/ColossalChat/rl_example.py [hunk 6]
```diff
@@ -227,13 +227,13 @@
...
     generate_config = dict(top_k=args.top_k, top_p=args.top_p, temperature=args.temperature)
 
     if args.backend == "transformers":
         inference_model_config.update(
             dict(
-                use_flash_attention_2=True,
+                use_flash_attention_2=False,
```
Comment (YeAnbang) [inline applications/ColossalChat/rl_example.py:236]: same here

File: applications/ColossalChat/rl_example.py
Comment (YeAnbang) [inline applications/ColossalChat/rl_example.py:1]: probably also consider force num_generation to 1 for reinforce++

File: applications/ColossalChat/coati/distributed/untitled.txt (hunk 4 above)
Comment (TongLi3701) [reply applications/ColossalChat/coati/distributed/untitled.txt:2]: Please remove this file.

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks, we left some comments.
